File: firebase/cloud-functions/functions/src/utils/ref.ts
import * as admin from "firebase-admin";
export class Ref {
  static get auth(): admin.auth.Auth {
    return admin.auth();
  }

  static get db(): admin.firestore.Firestore {
    return admin.firestore();
  }

  static get rdb(): admin.database.Database {
    return admin.database();
  }

  static get settingsCol() {
    return this.db.collection("settings");
  }

  static get adminDoc() {
    return this.settingsCol.doc("admins");
  }

  static get users() {
    return this.db.collection("users");
  }

  static userDoc(
    uid: string
  ): admin.firestore.DocumentReference<admin.firestore.DocumentData> {
    return this.users.doc(uid);
  }

  static get userSettings() {
    return this.db.collection("user_settings");
  }

  static userSettingDoc(
    uid: string
  ): admin.firestore.DocumentReference<admin.firestore.DocumentData> {
    return this.userSettings.doc(uid);
  }

  static get usersPublicDataCol() {
    return this.db.collection("users_public_data");
  }

  // Returns user public data document path
  static publicDoc(
    uid: string
  ): admin.firestore.DocumentReference<admin.firestore.DocumentData> {
    return this.usersPublicDataCol.doc(uid);
  }

  static get userSettingsCol(): admin.firestore.CollectionReference {
    return this.db.collection("user_settings");
  }

  static userSettingsDoc(uid: string): admin.firestore.DocumentReference {
    return this.userSettingsCol.doc(uid);
  }

  static get categoryCol() {
    return this.db.collection("categories");
  }

  /**
   * Returns category referrence
   *
   * @param {*} id Category id
   * @return reference
   */
  static categoryDoc(id: string) {
    return this.categoryCol.doc(id);
  }

  static get postCol() {
    return this.db.collection("posts");
  }

  /**
   * Returns post reference
   * @param id post id
   * @return reference
   */
  static postDoc(id: string) {
    return this.postCol.doc(id);
  }

  static get commentCol() {
    return this.db.collection("comments");
  }
  static commentDoc(commentId: string): admin.firestore.DocumentReference {
    return this.commentCol.doc(commentId);
  }

  static get chatRoomsCol() {
    return this.db.collection("chat_rooms");
  }

  static chatRoomsDoc(
    docId: string
  ): admin.firestore.DocumentReference<admin.firestore.DocumentData> {
    return this.chatRoomsCol.doc(docId);
  }
  static chatRoomDoc(
    docId: string
  ): admin.firestore.DocumentReference<admin.firestore.DocumentData> {
    return this.chatRoomsDoc(docId);
  }

  static get chatRoomMessagesCol() {
    return this.db.collection("chat_room_messages");
  }

  /** *************** MESSAGING References **************/
  static get tokenCollectionGroup() {
    return Ref.db.collectionGroup("fcm_tokens");
  }

  static tokenCol(uid: string): admin.firestore.CollectionReference {
    return this.userDoc(uid).collection("fcm_tokens");
  }
  static tokenDoc(
    uid: string,
    token: string
  ): admin.firestore.DocumentReference {
    return this.tokenCol(uid).doc(token);
  }

  /** Point */

  static pointHistoryCol(uid: string): admin.firestore.CollectionReference {
    return this.userDoc(uid).collection("point_history");
  }

  // Point history folder for post point events.
  static pointLastHistory(
    uid: string,
    eventName: string
  ): admin.firestore.Query<admin.firestore.DocumentData> {
    // console.log(this.pointHistoryCol(uid).path, eventName);
    return this.pointHistoryCol(uid)
      .where("eventName", "==", eventName)
      .orderBy("createdAt", "desc")
      .limit(1);
  }
}

File: lib/src/functions/ref.dart
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:firebase_auth/firebase_auth.dart';
import 'package:firebase_database/firebase_database.dart';
import 'package:fireflutter/fireflutter.dart';

const String userCollectionName = 'users';

/// Firestore database instance
FirebaseFirestore get db => FirebaseFirestore.instance;

/// Firebase Realtime Database instance
FirebaseDatabase get rtdb => FirebaseDatabase.instance;

/// user
CollectionReference get userCol =>
    FirebaseFirestore.instance.collection(userCollectionName);
DocumentReference userDoc(String uid) => userCol.doc(uid);
DocumentReference get myDoc => userDoc(FirebaseAuth.instance.currentUser!.uid);

CollectionReference get userPrivateCol =>
    FirebaseFirestore.instance.collection('user_private_data');
DocumentReference get myPrivateDoc => userPrivateCol.doc(myUid);

// categories
CollectionReference get categoryCol =>
    FirebaseFirestore.instance.collection(Category.collectionName);
DocumentReference categoryDoc(String categoryId) => categoryCol.doc(categoryId);

/// post
CollectionReference get postCol =>
    FirebaseFirestore.instance.collection('posts');
DocumentReference postDoc(String postId) => postCol.doc(postId);

CollectionReference get commentCol =>
    FirebaseFirestore.instance.collection('comments');
DocumentReference commentDoc(String postId) => postCol.doc(postId);

/// chat
CollectionReference get chatCol =>
    FirebaseFirestore.instance.collection('chats');
CollectionReference messageCol(String roomId) =>
    chatCol.doc(roomId).collection('messages');
DocumentReference roomRef(String roomId) => chatCol.doc(roomId);
DocumentReference roomDoc(String roomId) => chatCol.doc(roomId);

//
// DatabaseReference noOfNewMessageRef(String roomId) =>
//     rtdb.ref('chats/noOfNewMessages/');
//
DatabaseReference noOfNewMessageRef({required String uid}) =>
    rtdb.ref('chats/noOfNewMessages/$uid');

DocumentReference tokenDoc(String token) {
  return myDoc.collection('fcm_tokens').doc(token);
}

CollectionReference tokensCol(String uid) {
  return userCol.doc(uid).collection('fcm_tokens');
}

/// User setting
/// Note, for the sign-in user's setting, you should use `UserService.instance.settings`
/// Note, for other user settings, you should use `UserSettings(otherUid, docId)`.
CollectionReference userSettingCol(String uid) =>
    userDoc(uid).collection('user_settings');
CollectionReference get mySettingCol => userSettingCol(myUid!);

DocumentReference mySettingDoc(String documentid) =>
    mySettingCol.doc(documentid);

// favorites
CollectionReference get favoriteCol => db.collection('favorites');
DocumentReference favoriteDoc(String postId) => favoriteCol.doc(postId);

/// push notifications
CollectionReference get messageQueueCol =>
    db.collection('push_notification_queue');

/// Number of Profile View
CollectionReference get profileViewHistoryCol =>
    db.collection('profile_view_history');
DocumentReference profileViewHistoryDoc(
        {required String myUid, required String otherUid}) =>
    profileViewHistoryCol.doc('$myUid-$otherUid');

/// Returns a string of database path to the login user's block list.
///
/// [otherUid] is the uid of the user to be blocked or unblocked.
/// If [otherUid] is null or empty, it returns the path to the login user's block list.
String pathBlock(String? otherUid) {
  if (otherUid == null || otherUid.isEmpty) {
    return 'blocks/$myUid';
  } else {
    return 'blocks/$myUid/$otherUid';
  }
}

String pathSeenBy(String id) {
  return 'posts/$id/seenBy';
}

String pathPostLikedBy(String id, {bool all = false}) {
  if (all) {
    return 'posts/$id/likedBy';
  } else {
    return 'posts/$id/likedBy/$myUid';
  }
}

String pathCommentLikedBy(String id, {bool all = false}) {
  if (all) {
    return 'comments/$id/likedBy';
  } else {
    return 'comments/$id/likedBy/$myUid';
  }
}

String pathUserLiked(String id, {bool all = false}) {
  if (all) {
    return 'likes/$id';
  } else {
    return 'likes/$id/$myUid';
  }
}

/// For the path to the users' blocked list
String pathUserBlocked(String id, {bool all = false}) {
  if (all) {
    return 'blocks/$id';
  } else {
    return 'blocks/$id/$myUid';
  }
}

/// storages
CollectionReference get storageCol =>
    FirebaseFirestore.instance.collection('storages');
DocumentReference storageDoc(String storageId) => storageCol.doc(storageId);

/// reports
CollectionReference get reportCol =>
    FirebaseFirestore.instance.collection('reports');

DocumentReference reportDoc(String id) => reportCol.doc(id);

/// Activity logs
DatabaseReference get activityLogRef => rtdb.ref('activity_logs/');
